File: symbsat-cpp/zdd.cpp
#include "zdd.h"

// TODO need to review for correctness
bool ZDD::operator==(const ZDD& rhs) const {

    if (this == &rhs) {
        return true;
    }

    MonomConstIterator it1(*this), it2(rhs);

    while (!it1 || !it2) {
        if ( !(it1.monom() == it2.monom()) ) {
            return false;
        }
        ++it1;
        ++it2;
    }

    // TODO should be used || or ^ ?
    if (!it1 ^ !it2) {
        return false;
    }

    return true;
}

const ZDD::Node* ZDD::add(const Node* i, const Node* j) {
    if ( i->isZero() ) {
        return copy(j);
    } else if ( j->isZero() ) {
        return copy(i);
    // } else if (i == j) {
    } else if (ZDD::Node::isEqual(i, j)) {
        // return 0
        return mZero;
    } else if ( i->isOne() ) {
        return create_node(
            j->mVar, copy(j->mMul),
            add(j->mAdd, mOne)
        );
    } else if ( j->isOne() ) {
        return create_node(
            i->mVar, copy(i->mMul),
            add(i->mAdd, mOne)
        );
    } else {
        if (i->mVar < j->mVar) {
            return create_node(
                i->mVar, copy(i->mMul), add(i->mAdd, j)
            );
        } else if (i->mVar > j->mVar) {
            return create_node(
                // j->mVar, copy(j->mMul), add(j->mAdd, i)
                j->mVar, j->mMul, add(i, j->mAdd)
            );
        } else {
            auto m = add(i->mMul, j->mMul);
            auto a = add(i->mAdd, j->mAdd);

            if (m->isZero()) {
                return a;
            }

            return create_node(i->mVar, m, a);
        }
    }
}

const ZDD::Node* ZDD::mul(const Node* i, const Node* j) {
    if ( i->isOne() ) {
        return copy(j);
    } else if ( i->isZero() || j->isZero() ) {
        // return 0
        return mZero;
    // } else if ( j->isOne() || i == j ) {
    } else if ( j->isOne() || ZDD::Node::isEqual(i, j) ) {
        return copy(i);
    } else {
        if (i->mVar < j->mVar) {
            auto m = mul(i->mMul, j);
            auto a = mul(i->mAdd, j);

            if (m->isZero()) {
                return a;
            }

            return create_node(i->mVar, m, a);
        } else if (i->mVar > j->mVar) {
            auto m = mul(j->mMul, i);
            auto a = mul(j->mAdd, i);

            if (m->isZero()) {
                return a;
            }

            return create_node(j->mVar, m, a);
        } else {
            auto m1 = mul(i->mAdd, j->mMul);
            auto m2 = mul(i->mMul, j->mMul);
            auto m3 = mul(i->mMul, j->mAdd);
            auto m = add(m1, add(m2, m3));

            if (m->isZero()) {
                return mul(i->mAdd, j->mAdd);
            }

            return create_node(
                i->mVar, m,
                mul(i->mAdd, j->mAdd)
            );
        }
    }
}

ZDD ZDD::operator+(const ZDD& rhs) const {
    ZDD tmp(*this);
    tmp.mRoot = tmp.add(mRoot, rhs.mRoot);
    return tmp;
}

ZDD ZDD::operator*(const ZDD& rhs) const {
    ZDD tmp(*this);
    tmp.mRoot = tmp.mul(mRoot, rhs.mRoot);
    return tmp;
}

void ZDD::MonomConstIterator::operator++() {
    while(!mPath.empty() && (mPath.top()->mAdd->isZero())) {
        mPath.pop();
        mMonom.erase(std::begin(mMonom));
    }
    if (!mPath.empty()) {
        const Node *i=mPath.top()->mAdd;
        mPath.pop();
        mMonom.erase(std::begin(mMonom));
        for ( ; !i->isOne(); i = i->mMul) {
            mPath.push(i);
            mMonom.push_back(i->mVar);
        }
    }
}

std::ostream& operator<<(std::ostream& out, const ZDD::Node* a) {
    if (a->isZero()) {
        out << "_zero";
    } else if (a->isOne()) {
        out << "_one";
    } else {
        out << a->mVar << " -> {" << a->mMul << "} {" << a->mAdd << "} ";
    }

    return out;
}

std::ostream& operator<<(std::ostream& out, const ZDD &a) {
    ZDD::MonomConstIterator it(a);

    if (a.isZero()) {
        out << "0";
    } else if (a.isOne()) {
        out << "1";
    } else {
        while(!it) {
            out << it.monom() << " ";
            ++it;
        }
    }
    return out;
}

File: symbsat-py/symbsat/zdd_new/zdd.py
from .node import Node


class ZDD:

    def __init__(self, var=-1):
        if var < 0:
            self.root = Node.zero()
        else:
            self.root = self.create_node(
                var, Node.one(), Node.zero()
            )

    def create_node(self, var: int, m: Node, a: Node):
        return Node(var, m, a)

    @classmethod
    def zero(cls):
        return cls()

    @classmethod
    def one(cls):
        zdd = cls()
        zdd.root = Node.one()
        return zdd

    def _add(self, i, j):

        if i.is_zero():
            return j
        if j.is_zero():
            return i
        if i == j:
            return self.zero()

        if i.is_one():
            r = self.create_node(
                j.var, j.mul, self._add(j.add, Node.one())
            )
        elif j.is_one():
            r = self.create_node(
                i.var, i.mul, self._add(i.add, Node.one())
            )
        else:
            if i.var < j.var:
                r = self.create_node(i.var, i.mul, self._add(i.add, j))
            elif i.var > j.var:
                r = self.create_node(j.var, j.mul, self._add(i, j.add))
            else:
                m = self._add(i.mul, j.mul)
                a = self._add(i.add, j.add)

                if m.is_zero():
                    return a

                r = self.create_node(i.var, m, a)
        return r

    def _mul(self, i, j):

        if i.is_one():
            return j
        if i.is_zero() or j.is_zero():
            return Node.zero()
        if j.is_one() or i == j:
            return i

        r = None
        if i.var < j.var:
            m = self._mul(i.mul, j)
            a = self._mul(i.add, j)

            if m.is_zero():
                return a

            r = self.create_node(i.var, m, a)
        elif i.var > j.var:
            m = self._mul(j.mul, i)
            a = self._mul(j.add, i)

            if m.is_zero():
                return a

            r = self.create_node(j.var, m, a)
        else:
            m1 = self._mul(i.add, j.mul)
            m2 = self._mul(i.mul, j.mul)
            m3 = self._mul(i.mul, j.add)
            ms_sum = self._add(m1, self._add(m2, m3))

            if ms_sum.is_zero():
                return self._mul(i.add, j.add)

            r = self.create_node(i.var, ms_sum, self._mul(i.add, j.add))
        return r

    def __mul__(self, other):
        self_cls = type(self)
        zdd = self_cls()
        zdd.root = zdd._mul(self.root, other.root)
        return zdd

    def __add__(self, other):
        self_cls = type(self)
        zdd = self_cls()
        zdd.root = zdd._add(self.root, other.root)
        return zdd

